From: Liam Howlett <liam.howlett@oracle.com>
Subject: maple_tree: fix return from mas_prealloc()
Date: Fri, 17 Jun 2022 13:46:20 +0000

Record the error code before overwriting the node pointer and return the
error.

Link: https://lkml.kernel.org/r/20220617134609.1771611-1-Liam.Howlett@oracle.com
Fixes: 2ef6c13ee71e (maple_tree: make mas_prealloc() error checking more generic)
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Cc: Qian Cai <quic_qiancai@quicinc.com>
Cc: Yu Zhao <yuzhao@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/maple_tree.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/lib/maple_tree.c~maple-tree-add-new-data-structure-fix-10
+++ a/lib/maple_tree.c
@@ -5654,6 +5654,8 @@ void mas_store_prealloc(struct ma_state
  */
 int mas_preallocate(struct ma_state *mas, void *entry, gfp_t gfp)
 {
+	int ret;
+
 	mas_set_alloc_req(mas, 1 + mas_mt_height(mas) * 3);
 	mas_alloc_nodes(mas, gfp);
 	if (likely(!mas_is_err(mas)))
@@ -5661,8 +5663,9 @@ int mas_preallocate(struct ma_state *mas
 
 	mas_set_alloc_req(mas, 0);
 	mas_destroy(mas);
+	ret = xa_err(mas->node);
 	mas->node = MAS_START;
-	return xa_err(mas->node);
+	return ret;
 }
 
 /*
_
